### PR TITLE
MP 15-18 DPatch and TOG defense, multi-GPU support for attacks

### PR DESCRIPTION
**Pull request by the-anne**
This PR enables users of advsecurenet to run adversarial training for both DPatch and TOG attacks. It also adds support for running adversarial attacks on multiple GPUs. Furthermore, it extends the suite of object detectors to 2 more models, and it extends the suite of datasets to 1 more dataset.

Implemented functionalities:

- Implemented the object detection adversarial training logic (with support both for Python package interface and CLI).
- Integrated FasterRCNN as a ready-to-go object detector for running and evaluating the attacks, evaluations and adversarial trainings.
- Extended yolov5 functionality for adversarial trainings.
- Integrated an example HuggingFace object detector (CustomRTDetrModel).
- Integrated PascalVOC dataset into the package.
- Added support for multi-GPU attacks for DPatch and TOG. Extended mAP evaluator to take into account images from different GPUs when evaluating.
- Added example notebooks showcasing how to run the attacks on multiple GPUs, run adversarial training, and run mAP evaluation.
- Refactored the codebase: removed CustomYolov5ODWrapper and custom wrappers per model. Moved the shared logic into a shared ODWrapper class and individual logic into model-specific CustomModel files.
- Created tests for the new functionalities.

Resolved issues:
- https://github.com/advsecurenet/advsecurenet_mp/issues/11
- https://github.com/advsecurenet/advsecurenet_mp/issues/10
- https://github.com/advsecurenet/advsecurenet_mp/issues/8
- https://github.com/advsecurenet/advsecurenet_mp/issues/7

### CODE REVIEW

File: advsecurenet/datasets/PascalVOC/pascalvoc_dataset.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        # If already extracted, skip downloading/extracting
+        voc_dir = os.path.join(root, "VOCdevkit", f"VOC{year}")
+        if os.path.isdir(voc_dir):
+            logger.info(f"Found existing {voc_dir}, skipping mirror download.")
+            return
+        if str(year) == "2007":
+            urls = [
+                "https://data.brainchip.com/dataset-mirror/voc/VOCtrainval_06-Nov-2007.tar",
+                "https://data.brainchip.com/dataset-mirror/voc/VOCtest_06-Nov-2007.tar",
+            ]
```
Comment (fabienmorgan) [inline advsecurenet/datasets/PascalVOC/pascalvoc_dataset.py]: It would be cleaner if these data.brainchip.com urls would be declared as constant at the top of the file and then to reference the constant in here.

File: advsecurenet/shared/types/configs/attack_configs/attacker_config.py [hunk 2]
```diff
@@ -17,10 +17,10 @@ class AttackerConfig:
     Configuration class for the Attacker module.
     """
 
-    model: torch.nn.Module
     attack: AdversarialAttack
     dataloader: Union[DataLoader, DataLoaderConfig]
     device: DeviceConfig
+    model: Optional[torch.nn.Module] = None
```
Comment (fabienmorgan) [inline advsecurenet/shared/types/configs/attack_configs/attacker_config.py:23]: Why did the model get Optional with a default value of None? Doesn't a attack always need a model?

File: advsecurenet/computer_vision/object_detection/attacks/attacker/ddp_od_attacker.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+                collate_fn=getattr(dl, "collate_fn", None),
+                drop_last=getattr(dl, "drop_last", False),
+            )
+            self._sampler = sampler
+
+    def run_task(self):
+        if dist.is_initialized():
+            visible_devices = os.environ.get("CUDA_VISIBLE_DEVICES", "all")
+            current_device = (
+                torch.cuda.current_device() if torch.cuda.is_available() else "cpu"
```
Comment (fabienmorgan) [inline advsecurenet/computer_vision/object_detection/attacks/attacker/ddp_od_attacker.py:64]: Why is there a else "cpu" in the ddp code? DDP requries cuda and doesn't run on the cpu

File: advsecurenet/models/CustomODModels/CustomFasterRCNNModel.py [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+            # img_list = []
+            # for i in range(x.shape[0]):
+            #     im = x[i].to(self.device)
+            #     # Ensure float32 and normalize if it looks like [0, 255]
+            #     if im.dtype != torch.float32:
+            #         im = im.float()
+            #     if im.max() > 1.0:
+            #         im = im / 255.0
+            #     img_list.append(im)
+            # x = img_list
```
Comment (fabienmorgan) [inline advsecurenet/models/CustomODModels/CustomFasterRCNNModel.py]: Is this commented code still required?

File: advsecurenet/models/CustomODModels/CustomFasterRCNNModel.py [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+        images = images.to(self.device).float()
+        img_list = [images[i] for i in range(images.size(0))]
+        for t in img_list:
+            t.requires_grad_(True)
+        n = len(next(iter(targets.values()))) if targets else 0
+        targets = [{k: v[i] for k, v in targets.items()} for i in range(n)]
+        return img_list, targets
+
+    def calculate_loss(self, predictions, target_val):
+        # target_val = 0 - subtract, target_val = 1 - add
```
Comment (fabienmorgan) [inline advsecurenet/models/CustomODModels/CustomFasterRCNNModel.py]: Also here. Can this comment be removed?

File: advsecurenet/models/CustomODModels/CustomFasterRCNNModel.py [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+                    imgs.append(t)
+            else:
+                t = x.to(self.device).float()
+                if t.max() > 1:
+                    t = t / 255.0
+                imgs = [t]
+        x_tensor = torch.stack(imgs, dim=0)
+        if requires_grad:
+            x_tensor.requires_grad_(True)
+        # return [x_tensor[i] for i in range(x_tensor.shape[0])]
```
Comment (fabienmorgan) [inline advsecurenet/models/CustomODModels/CustomFasterRCNNModel.py]: Same here

File: cli/logic/defense/adversarial_training/adversarial_training_cli.py [hunk 7]
```diff
@@ -105,11 +155,34 @@ def _prepare_models(self) -> List[BaseModel]:
...
+                        else False
+                    ),
+                    num_workers=train_loader.num_workers,
+                    pin_memory=train_loader.pin_memory,
+                    drop_last=train_loader.drop_last,
+                    collate_fn=train_loader.collate_fn,
+                )
+        except Exception:
+            pass
+        # --------------- TODO - remove this hack before merging ------------------------
```
Comment (fabienmorgan) [inline cli/logic/defense/adversarial_training/adversarial_training_cli.py]: I think you should remove this

**Review by fabienmorgan — CHANGES_REQUESTED** [review]
The code looks good. There are some minor changes that I suggest

File: advsecurenet/models/CustomODModels/CustomFasterRCNNModel.py [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+    # Model specific methods / helpers:
+
+    def _empty_target_np(self):
+        return {
+            "boxes": np.empty((0, 4), dtype=np.float32),
+            "labels": np.empty((0,), dtype=np.int64),
+        }
+
+    def _align_targets_to_batch(self, y, batch_size: int):
+        """
```
Comment (Rockyph) [inline advsecurenet/models/CustomODModels/CustomFasterRCNNModel.py:291]: In `translate_labels()` + `_align_targets_to_batch()`: when missing labels are padded via `_empty_target_np()`, they’re converted to tensors later but always on CPU and default dtypes. This can lead to device/dtype mismatch if input labels were already CUDA tensors. Consider creating empty targets directly on `self.device` with correct dtype.

File: advsecurenet/models/od_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,304 @@
+import numpy as np
+import torch
+import torch.distributed as dist
+import torch.nn as nn
+
+
+class ODWrapper:
+    def __init__(
+        self,
```
Comment (Rockyph) [inline advsecurenet/models/od_wrapper.py:9]: In ODWrapper.__init__: You directly modify the wrapped model (e.g. setting model.device, input_shape, channels_first). This can cause side effects if the same model instance is reused elsewhere perhaps it might be safer to keep these as wrapper-level properties rather than mutating the model.

**Review by Rockyph — CHANGES_REQUESTED** [review]
Code runs well, i've left two small comments to consider. Good job!!

File: advsecurenet/datasets/PascalVOC/pascalvoc_dataset.py (hunk 1 above)
Comment (the-anne) [reply advsecurenet/datasets/PascalVOC/pascalvoc_dataset.py]: Good point, thanks!

File: advsecurenet/models/CustomODModels/CustomFasterRCNNModel.py (hunk 4 above)
Comment (the-anne) [reply advsecurenet/models/CustomODModels/CustomFasterRCNNModel.py]: Good point, I will remove it!

File: advsecurenet/computer_vision/object_detection/attacks/attacker/ddp_od_attacker.py (hunk 3 above)
Comment (the-anne) [reply advsecurenet/computer_vision/object_detection/attacks/attacker/ddp_od_attacker.py:64]: `torch.distributed` technically supports CPU backends (e.g., `gloo` on multi-CPU nodes), so it is a fallback value for environments without GPUs - but you are right, it is not the expected scenario

I would still keep this fallback though as I think it is a nicer solution than throwing an error - but let me know if you disagree

File: advsecurenet/shared/types/configs/attack_configs/attacker_config.py (hunk 2 above)
Comment (the-anne) [reply advsecurenet/shared/types/configs/attack_configs/attacker_config.py:23]: Yes, you are right, we always need a model to run the attack. It is sort of a tricky solution which I chose because I wanted to avoid creating exactly the same class as `AttackerConfig` but with one field less for the purpose of object detection.

The reason why it is now changed to be an Optional is to make it compatible with object detection, which needs an object detector wrapper instead of the model - and injects it _on the go_ in the OD CLI. So object detection reuses all other fields from this class, apart from the `model`, which is why it must be an Optional field now.

File: advsecurenet/models/od_wrapper.py (hunk 9 above)
Comment (the-anne) [reply advsecurenet/models/od_wrapper.py:9]: Good point! Because I still need to pass information from object detector to model though, I think I will define a `configure` method inside the base model class which is essentially a setter for those fields - and then the wrapper will only call `model.configure(...)`, so it will no longer directly dabble with model's parameters.

**Review by Rockyph — APPROVED** [review]
Thanks for adressing my comments. Great job :)

**Review by fabienmorgan — APPROVED** [review]
Great fixes
